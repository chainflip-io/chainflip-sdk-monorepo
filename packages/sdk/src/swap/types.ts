import { Chain, Asset, chainAssets } from '@/shared/enums';
import { QuoteQueryResponse } from '@/shared/schemas';

export interface ChainData {
  id: Chain;
  name: string;
  isMainnet: boolean;
}

export type AssetData = {
  [K in keyof typeof chainAssets]: {
    id: (typeof chainAssets)[K][number];
    chain: K;
    contractAddress: string;
    decimals: number;
    name: string;
    symbol: string;
    isMainnet: boolean;
  };
}[keyof typeof chainAssets];

interface ChainsAndAssets {
  srcChain: Chain;
  destChain: Chain;
  srcAsset: Asset;
  destAsset: Asset;
}

export interface QuoteRequest extends ChainsAndAssets {
  amount: string;
}

export interface QuoteResponse extends ChainsAndAssets {
  quote: QuoteQueryResponse;
}

export interface SwapRequest extends QuoteRequest {
  destAddress: string;
}

export interface SwapResponse {
  id: string;
  depositAddress: string;
}

export interface SwapStatusRequest {
  swapDepositChannelId: string;
}

type CommonStatusFields = {
  depositAddress: string;
  destAddress: string;
  srcAsset: Asset;
  destAsset: Asset;
  expectedDepositAmount: string;
};

export type SwapStatusResponse = CommonStatusFields &
  (
    | { state: 'AWAITING_DEPOSIT' }
    | {
        state: 'DEPOSIT_RECEIVED';
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
      }
    | {
        state: 'SWAP_EXECUTED';
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
        swapExecutedAt: number;
        swapExecutedBlockIndex: string;
      }
    | {
        state: 'EGRESS_SCHEDULED';
        egressAmount: string;
        egressScheduledAt: number;
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
        swapExecutedAt: number;
        swapExecutedBlockIndex: string;
      }
    | {
        state: 'COMPLETE';
        egressAmount: string;
        egressCompletedAt: number;
        egressCompletedBlockIndex: string;
        egressScheduledAt: number;
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
        swapExecutedAt: number;
        swapExecutedBlockIndex: string;
      }
  );
