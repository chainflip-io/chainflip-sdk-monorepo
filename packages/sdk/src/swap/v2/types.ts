import { AffiliateBroker, FillOrKillParams, SwapFee } from '@/shared/schemas';
import { SwapType } from '@/swap/client';
import { FailureMode } from '@/swap/utils/swap';
import { ChainsAndAssets } from '../types';

interface DcaParameters {
  numberOfChunks: number;
  chunkIntervalBlocks: string;
}
interface Ccm {
  message: string;
  gasBudget: string;
  cfParameters: string | undefined;
  ccmDepositReceivedBlockIndex: string | undefined;
}
interface Failure {
  failedAt: number;
  failedAtBlockIndex: string;
  mode: FailureMode;
  reason: {
    code: string;
    message: string;
  };
}
interface Boost {
  maxBoostFeeBps: number;
  effectiveBoostFeeBps: number | undefined;
  boostedAt: number | undefined;
  boostedBlockIndex: string | undefined;
  skippedAt: number | undefined;
  skippedBlockIndex: string | undefined;
}
interface DepositChannelFields {
  createdAt: number;
  brokerCommissionBps: number;
  depositAddress: string;
  srcChainExpiryBlock: string;
  estimatedExpiryTime: number;
  expectedDepositAmount: string;
  isExpired: boolean;
  openedThroughBackend: boolean;
  affiliateBrokers: AffiliateBroker[];
  fillOrKillParams: FillOrKillParams | undefined;
  dcaParams: DcaParameters | undefined;
}
interface DepositFields {
  amount: string | undefined;
  txRef: string | undefined;
  txConfirmations: number | undefined;
  receivedAt: number | undefined;
  receivedBlockIndex: string | undefined;
  failure: Failure | undefined;
}
interface SwapChunk {
  swapInputAmount: string;
  swapOutputAmount: string | undefined;
  scheduledAt: number;
  scheduledBlockIndex: string;
  executedAt: number | undefined;
  executedBlockIndex: string | undefined;
  retryCount: number;
  latestSwapRescheduledAt: number | undefined;
  latestSwapRescheduledBlockIndex: number | undefined;
  fees: SwapFee[];
  isDca: boolean;
}
interface SwapFields {
  totalInputAmountSwapped: string | undefined;
  totalOutputAmountSwapped: string | undefined;
  totalChunksExecuted: number;
  isDca: boolean;
  lastExecutedChunk: SwapChunk | undefined;
  currentChunk: SwapChunk;
  type: SwapType;
  fees: SwapFee[];
}
interface EgressFields {
  amount: string;
  scheduledAt: number;
  scheduledBlockIndex: string;
  sentTxRef: string | undefined;
  sentAt: number | undefined;
  sentAtBlockIndex: string | undefined;
  ignoredAmount: string | undefined;
  failure: Failure | undefined;
  failedAt: number | undefined;
  failedAtBlockIndex: string | undefined;
}
interface SwapStatusResponseCommonFields extends ChainsAndAssets {
  swapId: string;
  destAddress: string;
  ccm: Ccm | undefined;
  boost: Boost | undefined;
  estimatedDurationSeconds: number | null | undefined;
  srcChainRequiredBlockConfirmations: number | null;
}

interface VaultSwapCommonFields extends Exclude<SwapStatusResponseCommonFields, 'boost'> {}
interface ReceivingVaultSwap extends VaultSwapCommonFields {
  deposit: Pick<DepositFields, 'amount' | 'txRef'>;
}
interface SwappingVaultSwap extends ReceivingVaultSwap {
  swap:
    | Exclude<SwapChunk, 'latestSwapRescheduledAt' | 'latestSwapRescheduledBlockIndex'>
    | undefined;
  swapEgress: EgressFields | undefined;
}
interface SendingVaultSwap extends SwappingVaultSwap {
  swapEgress: EgressFields | undefined;
}

interface DepositChannel extends SwapStatusResponseCommonFields {
  depositChannel: DepositChannelFields;
}
interface Receiving extends DepositChannel {
  deposit: DepositFields;
}
interface Swapping extends Receiving {
  swap: SwapFields;
}
interface Sending extends Receiving {
  swap: SwapFields;
  swapEgress: EgressFields | undefined;
  refundEgress: EgressFields | undefined;
}
type SwapState =
  | ({
      state: 'RECEIVING';
    } & Receiving)
  | ({
      state: 'SWAPPING';
    } & Swapping)
  | ({
      state: 'SENDING';
    } & Sending)
  | ({
      state: 'SENT';
    } & Sending)
  | ({
      state: 'COMPLETED';
    } & Sending)
  | ({
      state: 'FAILED';
    } & Sending);

export type DepositAddressStatusResponseV2 = SwapState;
export type VaultSwapStatusResponseV2 = SwapState & SendingVaultSwap;
export type SwapStatusResponseV2 = DepositAddressStatusResponseV2 | VaultSwapStatusResponseV2;
