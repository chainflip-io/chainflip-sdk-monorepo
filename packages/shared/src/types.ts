export type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;

export type Required<T, K extends keyof T> = T & { [P in K]-?: T[P] };
