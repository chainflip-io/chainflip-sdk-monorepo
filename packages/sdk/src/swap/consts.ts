export enum ChainId {
  Ethereum = 1,
  Polkadot,
  Bitcoin,
}

export type { SupportedAsset as TokenSymbol } from '@/shared/enums';

// value to be replaced at build time with `envsubst` or similar
export const BACKEND_SERVICE_URL = '$BACKEND_SERVICE_URL';
