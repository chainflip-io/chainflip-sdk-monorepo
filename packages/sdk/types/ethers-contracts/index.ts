/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
export type { ERC20 } from "./ERC20";
export type { Vault } from "./Vault";
export * as factories from "./factories";
export { ERC20__factory } from "./factories/ERC20__factory";
export { Vault__factory } from "./factories/Vault__factory";
