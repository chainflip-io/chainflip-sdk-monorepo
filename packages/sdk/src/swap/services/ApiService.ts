import axios from 'axios';
import type { ChainflipNetwork } from '@/shared/enums';
import type {
  QuoteQueryParams,
  QuoteResponse,
  SwapRequestBody,
} from '@/shared/schemas';
import { ChainId } from '../consts';
import {
  bitcoin,
  polkadot,
  dot$,
  ethereum,
  btc$,
  ethereumTokens,
  testnetChains,
  testnetTokens,
} from '../mocks';
import {
  Chain,
  RouteRequest,
  RouteResponse,
  SwapRequest,
  SwapResponse,
  SwapStatusRequest,
  SwapStatusResponse,
  Token,
} from '../types';
import { isTestnet } from '../utils';

const getChains = async (network: ChainflipNetwork): Promise<Chain[]> => {
  if (isTestnet(network)) {
    return testnetChains([ethereum, polkadot, bitcoin]);
  }
  return [ethereum, polkadot, bitcoin];
};

const getPossibleDestinationChains = async (
  chainId: ChainId,
  network: ChainflipNetwork,
): Promise<Chain[]> => {
  if (isTestnet(network)) {
    if (chainId === ChainId.Ethereum) return testnetChains([polkadot, bitcoin]);
    if (chainId === ChainId.Polkadot) return testnetChains([ethereum, bitcoin]);
    if (chainId === ChainId.Bitcoin) return testnetChains([ethereum, polkadot]);
    throw new Error('received testnet flag but mainnet chainId');
  }

  if (chainId === ChainId.Ethereum) return [bitcoin, polkadot];
  if (chainId === ChainId.Polkadot) return [ethereum, bitcoin];
  if (chainId === ChainId.Bitcoin) return [ethereum, polkadot];
  throw new Error('received unknown chainId');
};

const getTokens = async (
  chainId: ChainId,
  network: ChainflipNetwork,
): Promise<Token[]> => {
  if (isTestnet(network)) {
    if (chainId === ChainId.Ethereum) return testnetTokens(ethereumTokens);
    if (chainId === ChainId.Polkadot) return testnetTokens([dot$]);
    if (chainId === ChainId.Bitcoin) return testnetTokens([btc$]);
    throw new Error('received testnet flag but mainnet chainId');
  }

  if (chainId === ChainId.Ethereum) return ethereumTokens;
  if (chainId === ChainId.Polkadot) return [dot$];
  if (chainId === ChainId.Bitcoin) return [btc$];
  throw new Error('received unknown chainId');
};

export type RequestOptions = {
  signal?: AbortSignal;
};

type BackendQuery<T, U> = (
  baseUrl: string,
  args: T,
  options: RequestOptions,
) => Promise<U>;

const getRoute: BackendQuery<RouteRequest, RouteResponse> = async (
  baseUrl,
  { amount, ...routeRequest },
  { signal },
) => {
  const params: QuoteQueryParams = {
    amount,
    srcAsset: routeRequest.srcTokenSymbol,
    destAsset: routeRequest.destTokenSymbol,
  };

  const queryParams = new URLSearchParams(params);

  const url = new URL(`/quote?${queryParams.toString()}`, baseUrl).toString();

  const { data } = await axios.get<QuoteResponse>(url, { signal });

  return { quote: data, ...routeRequest };
};

const requestDepositAddress: BackendQuery<SwapRequest, SwapResponse> = async (
  baseUrl,
  route,
  { signal },
) => {
  const body: SwapRequestBody = {
    destAddress: route.destAddress,
    srcAsset: route.srcTokenSymbol,
    destAsset: route.destTokenSymbol,
    expectedDepositAmount: route.expectedDepositAmount,
  };

  const url = new URL('/swaps', baseUrl).toString();

  const { data } = await axios.post<SwapResponse>(url, body, { signal });

  return data;
};

const getStatus: BackendQuery<SwapStatusRequest, SwapStatusResponse> = async (
  baseUrl,
  { swapDepositChannelId },
  { signal },
): Promise<SwapStatusResponse> => {
  const url = new URL(`/swaps/${swapDepositChannelId}`, baseUrl).toString();
  const { data } = await axios.get<SwapStatusResponse>(url, {
    signal,
  });
  return data;
};

export default {
  getChains,
  getPossibleDestinationChains,
  getRoute,
  getTokens,
  getStatus,
  requestDepositAddress,
};
