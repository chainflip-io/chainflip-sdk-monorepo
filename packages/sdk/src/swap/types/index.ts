import { Network, SupportedAsset } from '@/shared/enums';
import { QuoteResponse } from '@/shared/schemas';
import { ChainId, TokenSymbol } from '../consts';

export type { SDKOptions } from '../sdk';

export interface Chain {
  id: ChainId;
  name: Network;
  isMainnet: boolean;
}
interface ChainTokenMap {
  [ChainId.Ethereum]: 'ETH' | 'USDC' | 'FLIP';
  [ChainId.Bitcoin]: 'BTC';
  [ChainId.Polkadot]: 'DOT';
}

export type Token = {
  [K in keyof ChainTokenMap]: {
    chainId: K;
    contractAddress: string;
    decimals: number;
    name: string;
    symbol: ChainTokenMap[K];
    isMainnet: boolean;
  };
}[keyof ChainTokenMap];

interface Route {
  srcChainId: ChainId;
  destChainId: ChainId;
  srcTokenSymbol: TokenSymbol;
  destTokenSymbol: TokenSymbol;
  destAddress: string;
}

export interface RouteRequest extends Route {
  amount: string;
}

export interface RouteResponse extends Route {
  quote: QuoteResponse;
}

export interface SwapRequest extends Omit<RouteResponse, 'quote'> {
  expectedDepositAmount: string;
}

export interface SwapResponse {
  id: string;
  depositAddress: string;
}

export interface SwapStatusRequest {
  swapDepositChannelId: string;
}

type CommonStatusFields = {
  depositAddress: string;
  destAddress: string;
  srcAsset: SupportedAsset;
  destAsset: SupportedAsset;
  expectedDepositAmount: string;
};

export type SwapStatusResponse = CommonStatusFields &
  (
    | { state: 'AWAITING_DEPOSIT' }
    | {
        state: 'DEPOSIT_RECEIVED';
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
      }
    | {
        state: 'SWAP_EXECUTED';
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
        swapExecutedAt: number;
        swapExecutedBlockIndex: string;
      }
    | {
        state: 'EGRESS_SCHEDULED';
        egressAmount: string;
        egressScheduledAt: number;
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
        swapExecutedAt: number;
        swapExecutedBlockIndex: string;
      }
    | {
        state: 'COMPLETE';
        egressAmount: string;
        egressCompletedAt: number;
        egressCompletedBlockIndex: string;
        egressScheduledAt: number;
        depositAmount: string;
        depositReceivedAt: number;
        depositReceivedBlockIndex: string;
        swapExecutedAt: number;
        swapExecutedBlockIndex: string;
      }
  );
