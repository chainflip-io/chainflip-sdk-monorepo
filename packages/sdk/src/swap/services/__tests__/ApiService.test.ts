import axios from 'axios';
import { ChainId } from '../../consts';
import { RouteRequest } from '../../types';
import ApiService from '../ApiService';

jest.mock('axios', () => ({
  get: jest.fn(),
  post: jest.fn(),
}));

describe('ApiService', () => {
  const mockRoute = {
    amount: '10000',
    destAddress: '',
    srcChainId: ChainId.Bitcoin,
    srcTokenSymbol: 'BTC',
    destChainId: ChainId.Ethereum,
    destTokenSymbol: 'ETH',
  } satisfies RouteRequest;

  describe(ApiService.getChains, () => {
    it.each(['sisyphos', 'perseverance'] as const)(
      'gets testnet chains (%s)',
      async (network) => {
        expect(await ApiService.getChains(network)).toMatchSnapshot();
      },
    );

    it('gets mainnet chains', async () => {
      expect(await ApiService.getChains('mainnet')).toMatchSnapshot();
    });
  });

  describe.each(['sisyphos', 'perseverance'] as const)(
    `${ApiService.getTokens.name} (%s)`,
    (network) => {
      it.each([ChainId.Ethereum, ChainId.Bitcoin, ChainId.Polkadot])(
        'gets the correct tokens for testnets (%d)',
        async (chainId) => {
          expect(
            await ApiService.getTokens(chainId, network),
          ).toMatchSnapshot();
        },
      );
    },
  );

  describe(ApiService.getTokens, () => {
    it.each([ChainId.Ethereum, ChainId.Bitcoin, ChainId.Polkadot])(
      'gets the correct tokens for mainnets (%d)',
      async (chainId) => {
        expect(
          await ApiService.getTokens(chainId, 'mainnet'),
        ).toMatchSnapshot();
      },
    );
  });

  describe(ApiService.getRoute, () => {
    it('gets a route with a quote', async () => {
      const mockedGet = jest.mocked(axios.get);
      mockedGet.mockResolvedValueOnce({
        data: {
          id: 'string',
          intermediateAmount: '1',
          egressAmount: '2',
        },
      });

      const route = await ApiService.getRoute(
        'https://swapperoo.org',
        {
          amount: '10000',
          destAddress: '',
          srcChainId: ChainId.Bitcoin,
          srcTokenSymbol: 'BTC',
          destChainId: ChainId.Ethereum,
          destTokenSymbol: 'ETH',
        },
        {},
      );

      expect(route).toMatchSnapshot();
      expect(mockedGet.mock.lastCall).toMatchSnapshot();
    });

    it('passes the signal to axios', async () => {
      const mockedGet = jest.mocked(axios.get);
      mockedGet.mockResolvedValueOnce({
        data: {
          id: 'string',
          intermediateAmount: '1',
          egressAmount: '2',
        },
      });

      await ApiService.getRoute('https://swapperoo.org', mockRoute, {
        signal: new AbortController().signal,
      });

      expect(mockedGet.mock.lastCall?.[1]?.signal).not.toBeUndefined();
    });
  });

  describe(ApiService.requestDepositAddress, () => {
    it('executes the route and returns the data from the service', async () => {
      const mockedPost = jest.mocked(axios.post);
      const response = {
        id: 'new deposit channel id',
        depositAddress: '0xcafebabe',
      };
      mockedPost.mockResolvedValueOnce({ data: response });

      const depositChannel = await ApiService.requestDepositAddress(
        'https://swapperoo.org',
        {
          ...mockRoute,
          expectedDepositAmount: mockRoute.amount,
        },
        {},
      );
      expect(depositChannel).toEqual(response);
    });

    it('passes on the signal', async () => {
      const mockedPost = jest.mocked(axios.post);
      mockedPost.mockResolvedValueOnce({
        data: { id: 'new deposit channel id', depositAddress: '0xcafebabe' },
      });

      await ApiService.requestDepositAddress(
        'https://swapperoo.org',
        {
          ...mockRoute,
          expectedDepositAmount: mockRoute.amount,
        },
        {
          signal: new AbortController().signal,
        },
      );
      expect(mockedPost.mock.lastCall?.[2]?.signal).not.toBeUndefined();
    });
  });

  describe(ApiService.getStatus, () => {
    it('forwards whatever response it gets from the swap service', async () => {
      const mockedGet = jest.mocked(axios.get);
      mockedGet.mockResolvedValueOnce({ data: 'hello darkness' });
      mockedGet.mockResolvedValueOnce({ data: 'my old friend' });

      const statusRequest = { swapDepositChannelId: 'the id' };

      const status1 = await ApiService.getStatus(
        'https://swapperoo.org',
        statusRequest,
        {},
      );
      expect(status1).toBe('hello darkness');
      const status2 = await ApiService.getStatus(
        'https://swapperoo.org',
        statusRequest,
        {},
      );
      expect(status2).toBe('my old friend');
    });

    it('passes the signal to axios', async () => {
      const mockedGet = jest.mocked(axios.get);
      mockedGet.mockResolvedValueOnce({ data: null });

      await ApiService.getStatus(
        'https://swapperoo.org',
        { swapDepositChannelId: '' },
        { signal: new AbortController().signal },
      );

      expect(mockedGet.mock.lastCall?.[1]?.signal).not.toBeUndefined();
    });
  });
});
