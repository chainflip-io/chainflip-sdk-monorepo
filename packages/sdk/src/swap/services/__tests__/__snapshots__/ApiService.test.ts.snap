// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiService getAssets (backspin) gets the correct assets for networks (Bitcoin) 1`] = `
[
  {
    "asset": "BTC",
    "chain": "Bitcoin",
    "chainflipId": "Btc",
    "contractAddress": undefined,
    "decimals": 8,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "600",
    "minimumSwapAmount": "390000",
    "name": "Bitcoin",
    "symbol": "BTC",
  },
]
`;

exports[`ApiService getAssets (backspin) gets the correct assets for networks (Ethereum) 1`] = `
[
  {
    "asset": "ETH",
    "chain": "Ethereum",
    "chainflipId": "Eth",
    "contractAddress": undefined,
    "decimals": 18,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "580000000000000",
    "name": "Ether",
    "symbol": "ETH",
  },
  {
    "asset": "USDC",
    "chain": "Ethereum",
    "chainflipId": "Usdc",
    "contractAddress": "0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0",
    "decimals": 6,
    "isMainnet": false,
    "maximumSwapAmount": "1000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000",
    "name": "USDC",
    "symbol": "USDC",
  },
  {
    "asset": "FLIP",
    "chain": "Ethereum",
    "chainflipId": "Flip",
    "contractAddress": "0x10C6E9530F1C1AF873a391030a1D9E8ed0630D26",
    "decimals": 18,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000000000000000",
    "name": "FLIP",
    "symbol": "FLIP",
  },
]
`;

exports[`ApiService getAssets (backspin) gets the correct assets for networks (Polkadot) 1`] = `
[
  {
    "asset": "DOT",
    "chain": "Polkadot",
    "chainflipId": "Dot",
    "contractAddress": undefined,
    "decimals": 10,
    "isMainnet": false,
    "maximumSwapAmount": "2000000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "2000000000",
    "name": "Polkadot",
    "symbol": "DOT",
  },
]
`;

exports[`ApiService getAssets (mainnet) gets the correct assets for networks (Bitcoin) 1`] = `
[
  {
    "asset": "BTC",
    "chain": "Bitcoin",
    "chainflipId": "Btc",
    "contractAddress": undefined,
    "decimals": 8,
    "isMainnet": true,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "600",
    "minimumSwapAmount": "390000",
    "name": "Bitcoin",
    "symbol": "BTC",
  },
]
`;

exports[`ApiService getAssets (mainnet) gets the correct assets for networks (Ethereum) 1`] = `
[
  {
    "asset": "ETH",
    "chain": "Ethereum",
    "chainflipId": "Eth",
    "contractAddress": undefined,
    "decimals": 18,
    "isMainnet": true,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "580000000000000",
    "name": "Ether",
    "symbol": "ETH",
  },
  {
    "asset": "USDC",
    "chain": "Ethereum",
    "chainflipId": "Usdc",
    "contractAddress": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    "decimals": 6,
    "isMainnet": true,
    "maximumSwapAmount": "1000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000",
    "name": "USDC",
    "symbol": "USDC",
  },
  {
    "asset": "FLIP",
    "chain": "Ethereum",
    "chainflipId": "Flip",
    "contractAddress": "0x826180541412D574cf1336d22c0C0a287822678A",
    "decimals": 18,
    "isMainnet": true,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000000000000000",
    "name": "FLIP",
    "symbol": "FLIP",
  },
]
`;

exports[`ApiService getAssets (mainnet) gets the correct assets for networks (Polkadot) 1`] = `
[
  {
    "asset": "DOT",
    "chain": "Polkadot",
    "chainflipId": "Dot",
    "contractAddress": undefined,
    "decimals": 10,
    "isMainnet": true,
    "maximumSwapAmount": "2000000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "2000000000",
    "name": "Polkadot",
    "symbol": "DOT",
  },
]
`;

exports[`ApiService getAssets (perseverance) gets the correct assets for networks (Bitcoin) 1`] = `
[
  {
    "asset": "BTC",
    "chain": "Bitcoin",
    "chainflipId": "Btc",
    "contractAddress": undefined,
    "decimals": 8,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "600",
    "minimumSwapAmount": "390000",
    "name": "Bitcoin",
    "symbol": "BTC",
  },
]
`;

exports[`ApiService getAssets (perseverance) gets the correct assets for networks (Ethereum) 1`] = `
[
  {
    "asset": "ETH",
    "chain": "Ethereum",
    "chainflipId": "Eth",
    "contractAddress": undefined,
    "decimals": 18,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "580000000000000",
    "name": "Ether",
    "symbol": "ETH",
  },
  {
    "asset": "USDC",
    "chain": "Ethereum",
    "chainflipId": "Usdc",
    "contractAddress": "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "decimals": 6,
    "isMainnet": false,
    "maximumSwapAmount": "1000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000",
    "name": "USDC",
    "symbol": "USDC",
  },
  {
    "asset": "FLIP",
    "chain": "Ethereum",
    "chainflipId": "Flip",
    "contractAddress": "0x0485D65da68b2A6b48C3fA28D7CCAce196798B94",
    "decimals": 18,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000000000000000",
    "name": "FLIP",
    "symbol": "FLIP",
  },
]
`;

exports[`ApiService getAssets (perseverance) gets the correct assets for networks (Polkadot) 1`] = `
[
  {
    "asset": "DOT",
    "chain": "Polkadot",
    "chainflipId": "Dot",
    "contractAddress": undefined,
    "decimals": 10,
    "isMainnet": false,
    "maximumSwapAmount": "2000000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "2000000000",
    "name": "Polkadot",
    "symbol": "DOT",
  },
]
`;

exports[`ApiService getAssets (sisyphos) gets the correct assets for networks (Bitcoin) 1`] = `
[
  {
    "asset": "BTC",
    "chain": "Bitcoin",
    "chainflipId": "Btc",
    "contractAddress": undefined,
    "decimals": 8,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "600",
    "minimumSwapAmount": "390000",
    "name": "Bitcoin",
    "symbol": "BTC",
  },
]
`;

exports[`ApiService getAssets (sisyphos) gets the correct assets for networks (Ethereum) 1`] = `
[
  {
    "asset": "ETH",
    "chain": "Ethereum",
    "chainflipId": "Eth",
    "contractAddress": undefined,
    "decimals": 18,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "580000000000000",
    "name": "Ether",
    "symbol": "ETH",
  },
  {
    "asset": "USDC",
    "chain": "Ethereum",
    "chainflipId": "Usdc",
    "contractAddress": "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "decimals": 6,
    "isMainnet": false,
    "maximumSwapAmount": "1000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000",
    "name": "USDC",
    "symbol": "USDC",
  },
  {
    "asset": "FLIP",
    "chain": "Ethereum",
    "chainflipId": "Flip",
    "contractAddress": "0x2BbB561C6eaB74f358cA9e8a961E3A20CAE3D100",
    "decimals": 18,
    "isMainnet": false,
    "maximumSwapAmount": null,
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "1000000000000000000",
    "name": "FLIP",
    "symbol": "FLIP",
  },
]
`;

exports[`ApiService getAssets (sisyphos) gets the correct assets for networks (Polkadot) 1`] = `
[
  {
    "asset": "DOT",
    "chain": "Polkadot",
    "chainflipId": "Dot",
    "contractAddress": undefined,
    "decimals": 10,
    "isMainnet": false,
    "maximumSwapAmount": "2000000000",
    "minimumEgressAmount": "1",
    "minimumSwapAmount": "2000000000",
    "name": "Polkadot",
    "symbol": "DOT",
  },
]
`;

exports[`ApiService getChains gets mainnet chains 1`] = `
[
  {
    "chain": "Ethereum",
    "isMainnet": true,
    "name": "Ethereum",
  },
  {
    "chain": "Polkadot",
    "isMainnet": true,
    "name": "Polkadot",
  },
  {
    "chain": "Bitcoin",
    "isMainnet": true,
    "name": "Bitcoin",
  },
]
`;

exports[`ApiService getChains gets testnet chains (perseverance) 1`] = `
[
  {
    "chain": "Ethereum",
    "isMainnet": false,
    "name": "Ethereum",
  },
  {
    "chain": "Polkadot",
    "isMainnet": false,
    "name": "Polkadot",
  },
  {
    "chain": "Bitcoin",
    "isMainnet": false,
    "name": "Bitcoin",
  },
]
`;

exports[`ApiService getChains gets testnet chains (sisyphos) 1`] = `
[
  {
    "chain": "Ethereum",
    "isMainnet": false,
    "name": "Ethereum",
  },
  {
    "chain": "Polkadot",
    "isMainnet": false,
    "name": "Polkadot",
  },
  {
    "chain": "Bitcoin",
    "isMainnet": false,
    "name": "Bitcoin",
  },
]
`;

exports[`ApiService getQuote gets a quote 1`] = `
{
  "amount": "10000",
  "destAsset": "ETH",
  "destChain": "Ethereum",
  "quote": {
    "egressAmount": "2",
    "id": "string",
    "intermediateAmount": "1",
  },
  "srcAsset": "BTC",
  "srcChain": "Bitcoin",
}
`;

exports[`ApiService getQuote gets a quote 2`] = `
[
  "https://swapperoo.org/quote?amount=10000&srcChain=Bitcoin&srcAsset=BTC&destChain=Ethereum&destAsset=ETH",
  {
    "signal": undefined,
  },
]
`;

exports[`ApiService getQuote gets a quote with a broker commission 1`] = `
{
  "amount": "10000",
  "destAsset": "ETH",
  "destChain": "Ethereum",
  "quote": {
    "egressAmount": "2",
    "id": "string",
    "intermediateAmount": "1",
  },
  "srcAsset": "BTC",
  "srcChain": "Bitcoin",
}
`;

exports[`ApiService getQuote gets a quote with a broker commission 2`] = `
[
  "https://swapperoo.org/quote?amount=10000&srcChain=Bitcoin&srcAsset=BTC&destChain=Ethereum&destAsset=ETH&brokerCommissionBps=15",
  {
    "signal": undefined,
  },
]
`;
